src/helpers.rs

```diff
@@ -509,18 +555,83 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         Ok((true, string_length))
     }
 
+    /// Helper function to write an OsStr as a 0x0000-terminated u16-sequence, which is what
+    /// the Windows APIs usually handle. This function returns `Ok((false, length))` without trying
+    /// to write if `size` is not large enough to fit the contents of `os_string` plus a null
+    /// terminator. It returns `Ok((true, length))` if the writing process was successful. The
+    /// string length returned does not include the null terminator.
+    fn write_os_str_to_wide_str(
+        &mut self,
+        os_str: &OsStr,
+        mplace: MPlaceTy<'tcx, Tag>,
+        size: u64,
+    ) -> InterpResult<'tcx, (bool, u64)> {
+        #[cfg(target_os = "windows")]
+        fn os_str_to_u16vec(os_str: &OsStr) -> Vec<u16> {
+            std::os::windows::ffi::OsStrExt::encode_wide(os_str).collect()
+        }
+        #[cfg(not(target_os = "windows"))]
+        fn os_str_to_u16vec(os_str: &OsStr) -> Vec<u16> {
+            os_str.to_str().unwrap().encode_utf16().collect()
```

Comment:
This code here could panic if the user supplies a non-unicode string. That's a bug, user code should never be able to trigger a panic.

This should use `map_err` like the code above to raise a reasonable error message instead.